src/Microsoft.Graph.Core/Requests/AuthenticationHandler.cs

```diff
@@ -0,0 +1,120 @@
+        {
+
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="AuthenticationHandler"/>
+        /// <param name="authenticationProvider">An authentication provider to pass to <see cref="AuthenticationHandler"/> for authenticating requests.</param>
+        /// </summary>
+        public AuthenticationHandler(IAuthenticationProvider authenticationProvider)
+        {
+            AuthenticationProvider = authenticationProvider;
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="AuthenticationHandler"/>
+        /// </summary>
+        /// <param name="authenticationProvider">An authentication provider to pass to <see cref="AuthenticationHandler"/> for authenticating requests.</param>
+        /// <param name="innerHandler">A HTTP message handler to pass to the <see cref="AuthenticationHandler"/> for sending requests.</param>
+        public AuthenticationHandler(IAuthenticationProvider authenticationProvider, HttpMessageHandler innerHandler)
+        {
+            InnerHandler = innerHandler;
+            AuthenticationProvider = authenticationProvider;
+        }
+
+        /// <summary>
+        /// Check HTTP response message status code if it's unauthorized (401) or not
+        /// </summary>
+        /// <param name="httpResponseMessage">The <see cref="HttpResponseMessage"/>to send.</param>
+        /// <returns></returns>
+        private bool IsUnauthorized(HttpResponseMessage httpResponseMessage)
+        {
+            return httpResponseMessage.StatusCode == HttpStatusCode.Unauthorized;
+        }
+
+        /// <summary>
+        /// Retry sending HTTP request
+        /// </summary>
+        /// <param name="httpResponseMessage">The <see cref="HttpResponseMessage"/>to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>to send.</param>
+        /// <returns></returns>
+        private async Task<HttpResponseMessage> SendRetryAsync(HttpResponseMessage httpResponseMessage, CancellationToken cancellationToken)
+        {
+            int retryAttempt = 0;
+            while (retryAttempt < MaxRetry)
+            {
+                var originalRequest = httpResponseMessage.RequestMessage;
+
+                // Authenticate request using AuthenticationProvider
+                await AuthenticationProvider.AuthenticateRequestAsync(originalRequest);
+                httpResponseMessage = await base.SendAsync(originalRequest, cancellationToken);
+
+                retryAttempt++;
+
+                if (!IsUnauthorized(httpResponseMessage) || !ContentHelper.IsBuffered(originalRequest))
+                {
+                    // Re-issue the request to get a new access token
+                    return httpResponseMessage;
+                }
+            }
+
+            return httpResponseMessage;
+        }
+
+        /// <summary>
+        /// Send a HTTP request and retries the request when the response if unauthorized
+        /// </summary>
```

Comment:
We should probably add a note here to explain why we are retrying.  The only scenario that I can think of is because the token expired between Graph getting the request and the backend receiving the request.